### DISCUSSION TITLE
Whatever happened to incremental adoption?

### DISCUSSION THREAD

**Discussion by wjk** [microsoft/WindowsAppSDK#987] [General] [answered]
The [README for the Windows App SDK](https://github.com/microsoft/WindowsAppSDK/tree/main/docs#what-is-project-reunion) says:

> Your app can use the parts of Project Reunion that are most interesting to you without having to rewrite it into a new app model or deployment system. Project Reunion provides a single consistent API surface for both new and existing functionality that automatically adapts to your preferred app model and deployment mechanism.

All of these statements are blatantly false. For no apparent reason, the new features in the App SDK **cannot** be integrated into an existing project. While CoreWindow (and its associated lifecycle requirements) have since been deprecated, and developers now have the option of adopting the AppContainer instead of being forced to, there is still no support in WinUI 3 for integrating it into previously developed applications whatsoever since [WinUI 3 will not support XAML Islands](https://github.com/microsoft/microsoft-ui-xaml/blob/main/docs/roadmap.md). This rather important feature is currently listed under the “coming in some version, at some indeterminate point in the future, if we even decide to do it at all” column. Between this and the requirement that you use the WinUI Application class as your entry point if you want to use the WinUI Window class, you _must_ recreate your codebase from scratch if you want to use WinUI 3 _and_ move from whatever it is you’re currently using to WinUI 3 all at once. And that isn’t even mentioning the plethora of bugs and missing features in WinUI that make writing a quality app from scratch almost impossible.

Furthermore, while it is _technically_ possible to use MRT Core outside of a WinUI 3 app, doing so is not officially supported “at this time” and you still have to deal with MSIX package deployment issues (either through packaging your app in an MSIX package, or installing the one that is part of the redistributable installer). There is no support for true XCOPY deployment, or for situations where MSIX packages are blocked from installation for some reason (probably by the administrator).

This is a problem because ever since the initial release build of Windows 10 seven years ago, and now going into Windows 11, the only way for an app to be able to use all of the platform’s new shiny UI features is for it to use WinUI 3 (or its predecessor, UWP XAML). For some reason, these features are never backported to Win32/Windows Forms or WPF, despite Microsoft’s declarations to the contrary during the Windows 11 announcement. (Yes, while the UXTheme bitmaps are being updated to match the new light WinUI theme, proper support of Dark Mode still requires that you use WinUI 3 AFAIK.)

And to top it all off, [the lifecycle documentation](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/release-channels#servicing-and-support) states that support for older versions of Windows will be dropped aggressively, which makes it even harder to incrementally adopt WinUI 3 because you will then need to constantly rewrite your app to keep up with not-always-backwards-compatible changes to deployment systems and feature access techniques (e.g. MSIX Dynamic Dependencies).

**Comment by mdtauk** [comment]
I believe XAML Islands is a feature coming after 1.0, but its just not ready at launch.

**Comment by axelandrejs** [comment, marked as answer]
>> Furthermore, while it is technically possible to use MRT Core outside of a WinUI 3 app, doing so is not officially supported “at this time”

Can you check these samples and see if they help with your scenario?

https://github.com/microsoft/WindowsAppSDK-Samples/tree/main/MrtCore/wpf_packaged_app
https://github.com/microsoft/WindowsAppSDK-Samples/tree/main/MrtCore/console_unpackaged_app
https://github.com/microsoft/WindowsAppSDK-Samples/tree/main/MrtCore/winforms_unpackaged_app

**Reply by wjk** [reply]
Actually, that helps a lot. This is exactly what I needed. I didn’t think that this functionality was ready yet. Now I know. Thanks!

As an aside, does the bootstrap installer support running without elevation, and where can I find a precompiled version of it?

**Reply by axelandrejs** [reply]
Glad that helped! These came online pretty recently. @DrusTheAxe can you help with the bootstrap question?

**Reply by DrusTheAxe** [reply]
The Bootstrap API is available in the 0.8-preview build[1]. You can find information including links for more details in [this thread](https://github.com/microsoft/WindowsAppSDK/issues/1003#issuecomment-871908131).

[1] Despite the name the 0.8-preview is effectively an 'experimental' build in the new Channel lingo. The channels decision didn't reach closure in time for 0.8 to switch naming/process/strategy.

You should find the API header and lib for build-time in the NuGet.

For runtime you need the MSIX packages installed. You can find them in the NuGet, or you can run the installer and it'll install them for you. Go [here](https://docs.microsoft.com/windows/apps/windows-app-sdk/experimental-channel#deployment-for-unpackaged-apps-experimental-feature) to get a hold of the installer.

>does the bootstrap installer support running without elevation

Not sure if you're referring to the Bootstrap API provided via `Microsoft.ProjectReunion.Bootstrap.dll` or the installer provided via `ProjectReunionInstall.exe`. The former just updates your process space to make the <del>ProjectReunion</del>WindowsAppSDK runtime available for use; the latter installs the <del>ProjectReunion</del>WindowsAppSDK's MSIX packages for runtime use. If the latter...

Yes, the installer works without elevation. It's rather adaptive. It will install the packages for the current user and, if it's running elevated, _also_ provision the packages for all users. The latter requires admin privilege so it only happens if the installer's running elevated. But if you don't have the rights you'll still get the packages installed, just not provisioned so any other users on the machine will still need to install the packages (until/unless someone eventually provisions them).

There's [currently a bug](https://github.com/microsoft/WindowsAppSDK/issues/885) where it generates a non-fatal (harmless) access-denied message if it's not running elevated. It installs the packages for the current user (akin to [packageManager.AddPackageAsync()](https://docs.microsoft.com/uwp/api/windows.management.deployment.packagemanager.addpackageasync?view=winrt-20348) or [Add-AppxPackage](https://docs.microsoft.com/powershell/module/appx/add-appxpackage?view=windowsserver2019-ps)) and _also_ tries to provision the packages for all users, but that requires admin privilege. It should only be trying to provision packages if running elevated.
